sdk/lib/atl/atlconv.h

```diff
@@ -0,0 +1,314 @@
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = _wcsdup(psz);
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CA2WEX
+{
+public:
+    LPWSTR m_psz;
+    wchar_t m_szBuffer[t_nBufferLength];
+
+    CA2WEX(_In_z_ LPCSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CA2WEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CA2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CA2WEX is not copyable
+    CA2WEX(_In_ const CA2WEX&) throw() = delete;
+    CA2WEX& operator=(_In_ const CA2WEX&) throw() = delete;
+
+    void Init(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        int cchMax = MultiByteToWideChar(nCodePage, 0, psz, -1, NULL, 0);
+        if (cchMax <= t_nBufferLength)
+        {
+            MultiByteToWideChar(nCodePage, 0, psz, -1, m_szBuffer, _countof(m_szBuffer));
+            m_szBuffer[_countof(m_szBuffer) - 1] = 0;
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = (LPWSTR)malloc(cchMax * sizeof(WCHAR));
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+
+        MultiByteToWideChar(nCodePage, 0, psz, -1, m_psz, cchMax);
+        m_psz[cchMax - 1] = 0;
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CW2AEX
+{
+public:
+    LPSTR m_psz;
+    char m_szBuffer[t_nBufferLength];
+
+    CW2AEX(_In_z_ LPCWSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CW2AEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CW2AEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CW2AEX is not copyable
+    CW2AEX(_In_ const CW2AEX&) throw() = delete;
+    CW2AEX& operator=(_In_ const CW2AEX&) throw() = delete;
+
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        int cchMax = WideCharToMultiByte(nConvertCodePage, 0, psz, -1, NULL, 0, NULL, NULL);
+        if (cchMax <= t_nBufferLength)
+        {
+            WideCharToMultiByte(nConvertCodePage, 0, psz, -1, m_szBuffer, _countof(m_szBuffer),
+                                NULL, NULL);
+            m_szBuffer[_countof(m_szBuffer) - 1] = 0;
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = (LPSTR)malloc(cchMax * sizeof(CHAR));
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+
+        WideCharToMultiByte(nConvertCodePage, 0, psz, -1, m_psz, cchMax, NULL, NULL);
+        m_psz[cchMax - 1] = 0;
+    }
+};
+
+typedef CA2AEX<128> CA2A;
+typedef CW2AEX<128> CW2A;
+typedef CA2WEX<128> CA2W;
+typedef CW2WEX<128> CW2W;
+typedef CA2CAEX<128> CA2CA;
+typedef CW2CWEX<128> CW2CW;
```

Comment:
I *think* that if you drop the 128 here, it will use the default from the template.

Reply:
Fixed by using `using`.

Reply:
Used `typedef`.